mllib/src/main/scala/org/apache/spark/ml/feature/TargetEncoder.scala

```diff
@@ -0,0 +1,472 @@
+                    } else throw new SparkException(
+                      s"Values from column ${getLabelCol} must be binary (0,1) but got $label.")
+                  case TargetEncoder.TARGET_CONTINUOUS => // incremental mean
+                    // increment counter and iterate on mean for current & unseen categories
+                    agg(feature) +
+                      (category -> (1 + class_count,
+                        class_stat + ((label - class_stat) / (1 + class_count)))) +
+                      (TargetEncoder.UNSEEN_CATEGORY -> (1 + global_count,
+                        global_stat + ((label - global_stat) / (1 + global_count))))
+                }
+              }
+            }.toArray
+          } else agg  // ignore NaN-labeled observations
+        } else agg,  // ignore null-labeled observations
+
+        (agg1, agg2) => inputFeatures.indices.map {
+          feature => {
+            val categories = agg1(feature).keySet ++ agg2(feature).keySet
+            categories.map(category =>
+              category -> {
+                val (counter1, stat1) = agg1(feature).getOrElse(category, (0.0, 0.0))
+                val (counter2, stat2) = agg2(feature).getOrElse(category, (0.0, 0.0))
+                $(targetType) match {
+                  case TargetEncoder.TARGET_BINARY => (counter1 + counter2, stat1 + stat2)
+                  case TargetEncoder.TARGET_CONTINUOUS => (counter1 + counter2,
+                    ((counter1 * stat1) + (counter2 * stat2)) / (counter1 + counter2))
+                }
+              }).toMap
+          }
+        }.toArray)
+
+
+
+    val model = new TargetEncoderModel(uid, stats).setParent(this)
+    copyValues(model)
+  }
+
+  @Since("4.0.0")
+  override def copy(extra: ParamMap): TargetEncoder = defaultCopy(extra)
+}
+
+@Since("4.0.0")
+object TargetEncoder extends DefaultParamsReadable[TargetEncoder] {
+
+  // handleInvalid parameter values
+  private[feature] val KEEP_INVALID: String = "keep"
+  private[feature] val ERROR_INVALID: String = "error"
+  private[feature] val supportedHandleInvalids: Array[String] = Array(KEEP_INVALID, ERROR_INVALID)
+
+  // targetType parameter values
+  private[feature] val TARGET_BINARY: String = "binary"
+  private[feature] val TARGET_CONTINUOUS: String = "continuous"
+  private[feature] val supportedTargetTypes: Array[String] = Array(TARGET_BINARY, TARGET_CONTINUOUS)
+
+  private[feature] val UNSEEN_CATEGORY: Option[Double] = Some(-1)
+
+  @Since("4.0.0")
+  override def load(path: String): TargetEncoder = super.load(path)
+}
+
+/**
+ * @param stats  Array of statistics for each input feature.
+ *               Array( Map( Some(category), (counter, stat) ) )
+ */
+@Since("4.0.0")
+class TargetEncoderModel private[ml] (
+                     @Since("4.0.0") override val uid: String,
+                     @Since("4.0.0") val stats: Array[Map[Option[Double], (Double, Double)]])
```

Comment:
nit: is it possible to avoid the usage of `Option` in the model coefficient?

e.g.
```
val stats: Array[Map[Double, (Double, Double)]]), # for valid values;
val statForInvalid: (Double, Double), # for invalid values;
```

Reply:
None is reserved for null category
It's possible to avoid it by reserving another double value (i.e. -2)

Reply:
done